File: packages/synapse-sdk-py/src/synapse_sdk/transports/http.py
from __future__ import annotations

from dataclasses import asdict
from datetime import UTC, datetime
from email.utils import parsedate_to_datetime
import random
import time
from typing import Any, Sequence

import requests

from synapse_sdk.errors import SynapseTransportError
from synapse_sdk.types import Claim, ObservationEvent, RetryConfig


class HttpTransport:
    def __init__(
        self,
        api_url: str,
        api_key: str | None = None,
        *,
        retry: RetryConfig | None = None,
        session: requests.Session | None = None,
    ) -> None:
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key
        self.retry = retry or RetryConfig()
        self.session = session or requests.Session()

    def close(self) -> None:
        self.session.close()

    def __enter__(self) -> "HttpTransport":
        return self

    def __exit__(self, exc_type: object, exc: object, tb: object) -> None:
        self.close()

    def send_events(self, events: Sequence[ObservationEvent], *, idempotency_key: str | None = None) -> None:
        self._post_json(
            "/v1/events",
            {"events": [asdict(e) for e in events]},
            idempotency_key=idempotency_key,
        )

    def propose_fact(self, claim: Claim, *, idempotency_key: str | None = None) -> None:
        self._post_json(
            "/v1/facts/proposals",
            {"claim": asdict(claim)},
            idempotency_key=idempotency_key,
        )

    def ingest_memory_backfill(
        self,
        batch_payload: dict[str, Any],
        *,
        idempotency_key: str | None = None,
    ) -> None:
        self._post_json(
            "/v1/backfill/memory",
            batch_payload,
            idempotency_key=idempotency_key,
        )

    def _post_json(
        self,
        path: str,
        payload: dict[str, Any],
        *,
        idempotency_key: str | None = None,
    ) -> None:
        self.request_json(path, method="POST", payload=payload, idempotency_key=idempotency_key)

    def request_json(
        self,
        path: str,
        *,
        method: str = "GET",
        payload: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        method_upper = method.upper()
        attempt = 0
        while True:
            try:
                response = self.session.request(
                    method_upper,
                    f"{self.api_url}{path}",
                    json=payload if method_upper != "GET" else None,
                    params=params,
                    headers=self._headers(idempotency_key),
                    timeout=self.retry.timeout_seconds,
                )
            except requests.RequestException as exc:
                error = SynapseTransportError(
                    "Synapse API request failed due to network error",
                    retryable=True,
                )
                if not self._should_retry(attempt, error):
                    raise error from exc
                self._sleep_before_retry(attempt, retry_after_seconds=None)
                attempt += 1
                continue

            if 200 <= response.status_code < 300:
                if not response.text:
                    return {}
                try:
                    decoded = response.json()
                except ValueError:
                    return {}
                if isinstance(decoded, dict):
                    return decoded
                return {"data": decoded}

            retryable = response.status_code in self.retry.retryable_status_codes
            response_body = response.text[:2000] if response.text else None
            error = SynapseTransportError(
                f"Synapse API request failed with status {response.status_code}",
                retryable=retryable,
                status_code=response.status_code,
                response_body=response_body,
            )
            if not self._should_retry(attempt, error):
                raise error

            self._sleep_before_retry(attempt, retry_after_seconds=self._parse_retry_after(response))
            attempt += 1

    def _should_retry(self, attempt: int, error: SynapseTransportError) -> bool:
        return error.retryable and attempt < self.retry.max_retries

    def _sleep_before_retry(self, attempt: int, *, retry_after_seconds: float | None) -> None:
        exp_delay = min(self.retry.max_delay_seconds, self.retry.base_delay_seconds * (2**attempt))
        jitter_amplitude = max(0.0, exp_delay * self.retry.jitter_ratio)
        jitter = random.uniform(-jitter_amplitude, jitter_amplitude) if jitter_amplitude > 0 else 0.0
        delay = max(0.0, exp_delay + jitter)
        if retry_after_seconds is not None:
            delay = max(delay, retry_after_seconds)
        time.sleep(delay)

    def _parse_retry_after(self, response: requests.Response) -> float | None:
        raw_value = response.headers.get("Retry-After")
        if not raw_value:
            return None
        stripped = raw_value.strip()
        if stripped.isdigit():
            return float(max(0, int(stripped)))
        try:
            retry_dt = parsedate_to_datetime(stripped)
        except (TypeError, ValueError):
            return None
        if retry_dt.tzinfo is None:
            retry_dt = retry_dt.replace(tzinfo=UTC)
        return max(0.0, (retry_dt - datetime.now(UTC)).total_seconds())

    def _headers(self, idempotency_key: str | None = None) -> dict[str, str]:
        headers = {"content-type": "application/json"}
        if self.api_key:
            headers["authorization"] = f"Bearer {self.api_key}"
        if idempotency_key:
            headers["Idempotency-Key"] = idempotency_key
        return headers

File: packages/synapse-sdk-ts/src/transports/http.ts
import type { Claim, ObservationEvent, RequestOptions, RetryPolicyConfig, SynapseTransport } from "../types.js";

const DEFAULT_RETRYABLE_STATUS_CODES = [408, 409, 425, 429, 500, 502, 503, 504];

const DEFAULT_RETRY_POLICY: Required<RetryPolicyConfig> = {
  maxRetries: 3,
  baseDelayMs: 250,
  maxDelayMs: 5_000,
  jitterRatio: 0.2,
  timeoutMs: 10_000,
  retryableStatusCodes: DEFAULT_RETRYABLE_STATUS_CODES
};

export class SynapseTransportError extends Error {
  readonly statusCode?: number;
  readonly retryable: boolean;
  readonly responseBody?: string;

  constructor(
    message: string,
    options: {
      statusCode?: number;
      retryable: boolean;
      responseBody?: string;
      cause?: unknown;
    }
  ) {
    super(message, { cause: options.cause });
    this.name = "SynapseTransportError";
    this.statusCode = options.statusCode;
    this.retryable = options.retryable;
    this.responseBody = options.responseBody;
  }
}

export class HttpTransport implements SynapseTransport {
  private readonly apiUrl: string;
  private readonly apiKey?: string;
  private readonly retryPolicy: Required<RetryPolicyConfig>;

  constructor(apiUrl: string, apiKey?: string, retry?: RetryPolicyConfig) {
    this.apiUrl = apiUrl.replace(/\/+$/, "");
    this.apiKey = apiKey;
    this.retryPolicy = {
      ...DEFAULT_RETRY_POLICY,
      ...retry,
      retryableStatusCodes: retry?.retryableStatusCodes ?? DEFAULT_RETRYABLE_STATUS_CODES
    };
  }

  async sendEvents(events: ObservationEvent[], options?: RequestOptions): Promise<void> {
    await this.requestJson("/v1/events", { method: "POST", payload: { events }, idempotencyKey: options?.idempotencyKey });
  }

  async proposeFact(claim: Claim, options?: RequestOptions): Promise<void> {
    await this.requestJson("/v1/facts/proposals", { method: "POST", payload: { claim }, idempotencyKey: options?.idempotencyKey });
  }

  async ingestMemoryBackfill(payload: Record<string, unknown>, options?: RequestOptions): Promise<void> {
    await this.requestJson("/v1/backfill/memory", { method: "POST", payload, idempotencyKey: options?.idempotencyKey });
  }

  async requestJson<T = Record<string, unknown>>(
    path: string,
    options: {
      method?: "GET" | "POST";
      payload?: unknown;
      params?: Record<string, string | number | boolean | null | undefined>;
      idempotencyKey?: string;
    } = {}
  ): Promise<T> {
    const response = await this.requestWithRetry(path, {
      method: options.method ?? "GET",
      payload: options.payload,
      params: options.params,
      idempotencyKey: options.idempotencyKey
    });
    const raw = await safeReadBody(response);
    if (!raw) {
      return {} as T;
    }
    try {
      return JSON.parse(raw) as T;
    } catch {
      return {} as T;
    }
  }

  private async requestWithRetry(
    path: string,
    options: {
      method: "GET" | "POST";
      payload?: unknown;
      params?: Record<string, string | number | boolean | null | undefined>;
      idempotencyKey?: string;
    }
  ): Promise<Response> {
    for (let attempt = 0; attempt <= this.retryPolicy.maxRetries; attempt += 1) {
      try {
        const response = await this.fetchWithTimeout(path, options);
        if (response.ok) {
          return response;
        }

        const retryable = this.retryPolicy.retryableStatusCodes.includes(response.status);
        const bodyText = await safeReadBody(response);
        const error = new SynapseTransportError(
          `Synapse API request failed with status ${response.status}`,
          {
            statusCode: response.status,
            retryable,
            responseBody: bodyText
          }
        );

        if (!retryable || attempt >= this.retryPolicy.maxRetries) {
          throw error;
        }

        const retryAfterMs = parseRetryAfterMs(response.headers.get("retry-after"));
        const backoffMs = computeBackoffDelayMs(this.retryPolicy, attempt);
        await sleep(Math.max(retryAfterMs ?? 0, backoffMs));
      } catch (error) {
        const transportError = normalizeNetworkError(error);
        if (!transportError.retryable || attempt >= this.retryPolicy.maxRetries) {
          throw transportError;
        }
        await sleep(computeBackoffDelayMs(this.retryPolicy, attempt));
      }
    }
    throw new SynapseTransportError("Synapse API request failed after retries", { retryable: false });
  }

  private async fetchWithTimeout(
    path: string,
    options: {
      method: "GET" | "POST";
      payload?: unknown;
      params?: Record<string, string | number | boolean | null | undefined>;
      idempotencyKey?: string;
    }
  ): Promise<Response> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.retryPolicy.timeoutMs);
    try {
      const query = buildQueryString(options.params);
      const url = `${this.apiUrl}${path}${query}`;
      return await fetch(url, {
        method: options.method,
        headers: this.headers(options.idempotencyKey),
        body: options.method === "GET" ? undefined : JSON.stringify(options.payload ?? {}),
        signal: controller.signal
      });
    } finally {
      clearTimeout(timeout);
    }
  }

  private headers(idempotencyKey?: string): Record<string, string> {
    const headers: Record<string, string> = { "content-type": "application/json" };
    if (this.apiKey) {
      headers.authorization = `Bearer ${this.apiKey}`;
    }
    if (idempotencyKey) {
      headers["Idempotency-Key"] = idempotencyKey;
    }
    return headers;
  }
}

function normalizeNetworkError(error: unknown): SynapseTransportError {
  if (error instanceof SynapseTransportError) {
    return error;
  }
  const isTimeout = typeof DOMException !== "undefined" && error instanceof DOMException && error.name === "AbortError";
  return new SynapseTransportError(
    isTimeout ? "Synapse API request timed out" : "Synapse API request failed due to network error",
    { retryable: true, cause: error }
  );
}

function computeBackoffDelayMs(policy: Required<RetryPolicyConfig>, attempt: number): number {
  const expDelay = Math.min(policy.maxDelayMs, policy.baseDelayMs * 2 ** attempt);
  if (policy.jitterRatio <= 0) {
    return expDelay;
  }
  const jitterAmplitude = expDelay * policy.jitterRatio;
  const jitter = (Math.random() * 2 - 1) * jitterAmplitude;
  return Math.max(0, Math.round(expDelay + jitter));
}

function parseRetryAfterMs(retryAfterHeader: string | null): number | null {
  if (!retryAfterHeader) {
    return null;
  }
  const seconds = Number.parseInt(retryAfterHeader, 10);
  if (Number.isFinite(seconds)) {
    return Math.max(0, seconds * 1_000);
  }
  const retryAtMs = Date.parse(retryAfterHeader);
  if (Number.isNaN(retryAtMs)) {
    return null;
  }
  return Math.max(0, retryAtMs - Date.now());
}

async function safeReadBody(response: Response): Promise<string | undefined> {
  try {
    const body = await response.text();
    return body.slice(0, 2_000);
  } catch {
    return undefined;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function buildQueryString(params?: Record<string, string | number | boolean | null | undefined>): string {
  if (!params) {
    return "";
  }
  const query = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === null || value === undefined) continue;
    query.set(key, String(value));
  }
  const raw = query.toString();
  return raw ? `?${raw}` : "";
}
